rules/active_ruleset.go

```diff
@@ -618,6 +618,31 @@ func filteredPipelinesWithAggregationType(
 	return pipelines[:cur]
 }
 
+// mergeResultsForExistingID merges the next staged metadata into the current list of staged
+// metadatas while ensuring the cutover times of the staged metadatas are non-decreasing. This
+// is needed because the cutover times of staged metadata results produced by mapping rule matching
+// may not always be in ascending order. For example, if at time T0 a metric matches against a
+// mapping rule, and the filter of such rule changed at T1 such that the metric no longer matches
+// the rule, this would indicate the staged metadata at T0 would have a cutover time of T0,
+// whereas the staged metadata at T1 would have a cutover time of 0 (due to no rule match),
```

Comment:
should the deletion of rule at T1 create a staged metadata with cutover time at T1 instead of overriding here?

Reply:
Talked offline, this approach is chosen in that it keeps the mapping rule matching logic simple (it doesn't need to know previous cutover times etc), nor do we need to thread the previous cutover times to mapping rule matching logic and rollup rule matching logic, which is cleaner. Since `forwardMatchAt` is only called internally, this is an acceptable tradeoff.